### ISSUE TITLE
SLHA-LR1: preregister top-16 ranking-preserving projection requalification

### ISSUE THREAD

**Issue by CHECKUPAUTO** [Memorithm/SLHAv2#107]
## Objective

Requalify the **already-implemented** ranking-aware projection path from #64 on the current TinyStories 15M / llama.cpp physical integration, without inventing a new scorer and without using protected-holdout outcomes for training, model selection, layer selection, or hyperparameter tuning.

This is a bounded follow-up to the non-protected diagnostics #105 and #106.

## Why this candidate, and why only one

The current non-protected evidence says:

- #105: restoring a broader baseline top-set/order is strongly diagnostic, while a top-1-only intervention is insufficient and exact baseline Q·K at inference is non-deployable;
- #106: layer-localization effects are strongly non-additive, so there is no justified fixed "bad layer" subset to tune around.

The repository already contains a deployable-compatible training mechanism from #64:

- `Objective::PairwiseTopK` trains the projection to preserve baseline top-ranked keys;
- baseline logits are offline training labels only;
- the rank dataset records query/key material so candidates can be scored offline;
- the external-K deployable path explicitly forbids the ranking-label/oracle collection knobs.

Therefore this issue freezes **one** candidate rather than opening a new objective/hyperparameter search.

## Frozen candidate LR1

Candidate id: `slha-lr1-pairwise-top16-all-layers-v1`

Training objective and scope:

```text
objective        = pairwise-topk
top_k            = 16
layers           = all
margin           = 1.0
max_negatives    = 8
negative_policy  = boundary-then-seeded
seed             = 7
learning_rate    = 1e-9
epochs            = 2
steps             = 0
batch             = 16
max_keys          = 256
geometry_weight   = 0.25
codec              = mixed
```

Rules:

- no `top_k` sweep;
- no layer-subset sweep;
- no alternate objective (`hybrid`, `listwise`, etc.) in LR1;
- no learning-rate / epoch / margin / negative-count sweep;
- warm-start from the exact current mixed projection artifact used by the qualifying run, frozen by manifest/hash before training;
- if the existing trainer/versioned binding cannot represent the current TinyStories dataset safely, create a **new versioned binding** rather than weakening or rewriting the historical frozen binding.

`top_k=16` is fixed from the already-observed **non-protected** #105 diagnostic direction. It is not to be changed after implementation starts.

All six layers remain in scope because #106 does not support a stable additive layer-selection rule.

## Data separation

LR1 must have three distinct identities before any training run:

1. **training corpus / rank dataset** — non-protected;
2. **validation split** — non-protected and disjoint from training by deterministic chunk identity;
3. existing **diagnostic corpus** `tinystories_rank_diagnostic_v1.txt` — non-protected, used only after the candidate artifact is frozen for the run.

All three must be hash-bound.

The protected synthetic holdout (`tinystories_synthetic_holdout.txt`) is forbidden during LR1 development/selection.

The training/validation corpus identity and chunk split must be committed/frozen **before** optimizer execution. Do not derive the split from observed metric values.

## Required development evidence

Before any protected-holdout evaluation, LR1 must produce machine-readable evidence for the fixed candidate versus the unchanged current L2/mixed control:

- exact model/tokenizer/llama.cpp revision and hashes;
- training dataset aggregate hash;
- deterministic training and validation chunk identities;
- initial-weight manifest/hash;
- trainer binary/source binding;
- final candidate weight manifest/hash;
- top-16 set recall/overlap on validation;
- top-16 boundary pair accuracy or equivalent already-supported pairwise ranking metric;
- score-geometry diagnostic;
- non-protected diagnostic-corpus PPL using the same model/tokenizer/corpus/context/threads policy in both arms;
- strict replacement coverage/counters with zero fallback/failure;
- confirmation that inference did not read baseline logits or enable `SLHA_SCORE_ORACLE`, `SLHA_RANK_DATASET_DIR`, or other training/oracle knobs.

Ranking metrics are mechanistic development evidence, not substitutes for task quality.

### Stage A — non-protected development/qualification

Only the frozen LR1 candidate above is trained and compared with the unchanged control. No protected holdout is accessed.

If LR1 does not improve the intended ranking metric and the non-protected task-quality direction, record the negative result and stop. Do not rescue it by changing `k`, layers, objective, or hyperparameters under this candidate id.

### Stage B — protected final evaluation

A separate commit/PR must first freeze:

- the exact LR1 weight artifact hash;
- the exact final evaluation command and environment;
- the primary task-quality decision rule;
- all secondary memory/tokens-per-second reporting requirements;
- the no-retuning rule.

Only then may the protected holdout be run. The protected result validates or rejects the already-frozen LR1 artifact; it must never feed back into candidate tuning.

## Scientific / product boundaries

This issue does **not** claim that top-16 ranking preservation is sufficient, optimal, novel, or better than standard attention.

It does not authorize:

- baseline Q·K oracle access in deployment;
- protected-holdout tuning;
- per-layer policy tuning from #106;
- more WARM/COLD/CCOS complexity as a response to the quality gap;
- performance or memory claims from tile size alone;
- FLAT/ElasticXxx promotion.

The outcome may be negative. Executed evidence is authoritative.

## Definition of done for the first PR

The first PR is complete when it adds the frozen LR1 protocol/bindings and a reproducible **non-protected** Stage-A runner, passes all existing repository CI on the exact head, and does not execute the protected holdout.

**Comment by CHECKUPAUTO** [comment]
### Stage-A audit note — validation must be independent

Audit of the existing `scirust/src/bin/slha_train_rank.rs` after preregistering LR1 found an important implementation boundary: the current `--validation-manifest` output is only a copy of the training manifest; it does **not** execute the frozen candidate on held-out chunks or compute top-k/ranking validation metrics.

Therefore LR1 Stage A must **not** cite that field as validation evidence. The follow-up implementation needs a separate read-only evaluator that loads the frozen candidate weights and only the preregistered validation chunks (12–15), computes the required top-16 ranking/geometry diagnostics, and writes a distinct machine-readable validation artifact. It must not update weights, recompute training normalisers, select hyperparameters, or access the protected synthetic holdout.

This does not change the frozen LR1 candidate or split. It tightens the evidence boundary identified by code audit.

**Comment by CHECKUPAUTO** [comment]
### Structural preregistration correction — rank-dataset chunk indexing

Before any LR1 optimizer execution, the pinned `llama.cpp` perplexity control flow and the SLHA collector lifecycle were audited together.

The frozen b9860 perplexity loop calls `llama_memory_clear(llama_get_memory(ctx), true)` **before decoding each evaluation chunk**. The SLHA patch hooks that call to `slha_k_clear_all()`. When rank collection is enabled, `slha_k_clear_all()` seals the previous raw-K matrix and then calls `slha_rank_dataset::begin_chunk()`, which increments `State.chunk` from its initial value 0.

Therefore, for a 16-chunk collection:

- storage slot/chunk 0 is the pre-first-decode empty slot;
- populated evaluation rows are tagged with chunk ids **1..16**;
- the rank dataset key vector has 17 storage slots (0..16), with slot 0 empty;
- preserving the preregistered 12/4 chronological split requires **training chunks 1..12** and **validation chunks 13..16**.

The previously recorded `[0..11]` / `[12..15]` ids were an off-by-one interpretation of collector storage ids. They must be corrected before training. This correction is structural only: no LR1 training, validation, diagnostic-PPL or protected-holdout metric has been observed or used to make it.

PR #109 will update the machine-readable contract/checks and hard-coded trainer/evaluator splits consistently, and will fail closed if the expected empty slot 0 / populated 1..16 layout is not present. The candidate objective, top-k, layers and all optimizer hyperparameters remain unchanged.

**Comment by CHECKUPAUTO** [comment]
### Structural clarification — short-context rows

A pre-execution audit of the existing #64 `PairwiseTopK` implementation found another protocol detail that must be handled explicitly before LR1 runs.

The rank collector samples `t=0` (default token stride includes zero), so some rows have `n_visible <= 16`. In the existing objective:

- `PairwiseTopK(k=16)` naturally emits zero pairwise comparisons when the visible set has no key outside the top-16 boundary;
- the separately preregistered geometry penalty still applies on rows with at least two scores.

Therefore LR1 training will **retain** these rows, preserving the already-implemented #64 objective rather than inventing a new row filter.

For mechanistic validation, however, top-16 set recall and top-16-vs-rest pair accuracy are only informative when `n_visible > 16`. The Stage-A evaluator will compute those metrics only on that eligible subset, while reporting the eligible-row denominator explicitly. Spearman/score-geometry diagnostics use their own non-degenerate row counts. Short rows must not contribute trivial 1.0 top-16 overlap.

This clarification is based only on objective/collector semantics, before LR1 optimizer execution or any Stage-A outcome.
